Use Graphviz_jll

fingolfin opened the pull request:
This removes the need to manually install Graphviz

Resolves #16

giordano commented on src/PProf.jl:
```diff
@@ -326,7 +327,9 @@ function refresh(; webhost::AbstractString = "localhost",
     relative_percentages_flag = ui_relative_percentages ? "-relative_percentages" : ""
 
     proc[] = pprof_jll.pprof() do pprof_path
-        open(pipeline(`$pprof_path -http=$webhost:$webport $relative_percentages_flag $file`))
+        Graphviz_jll.dot() do dot_path
+            open(pipeline(`$pprof_path -http=$webhost:$webport $relative_percentages_flag $file`))
+        end
     end
```
Since this package requires Julia v1.6, you could use the [recommended thread-safe syntax](https://docs.binarybuilder.org/stable/jll/#ExecutableProduct)
```julia
run(`$(mungify_exe()) $arguments`)
```
instead of
```julia
mungify_exe() do exe
    run(`$exe $arguments`)
end
```
but I don't understand what this is doing: `dot_path` isn't used.  Is it used only to merge the environments?

vchuravy replied:
Yeah we just need to instead dot into the path, since it is going to be called by pprof

giordano replied:
Bit convoluted, but can do something like
```julia
pprof = addenv(pprof_jll.pprof(), Graphviz_jll.JLLWrappers.LIBPATH_env=>Graphviz_jll.LIBPATH[], "PATH"=>Graphviz_jll.PATH[])
open(pipeline(`$pprof -http=$webhost:$webport $relative_percentages_flag $file`))
```

fingolfin replied:
It is *not* enough to just modify PATH, because shared libs used by `dot` must also be found. 

I did the change the way I did it because it was the easiest and quickest way for me to achieve what I needed. I was not aware of the new syntax. I simply extended what was there (in particular it seems to me that any thread safety issues are already present and not just introduced by this PR?)

the patch by @giordano seems interesting, but I don't understand it right now -- e.g. shouldn't the "LIBPATH" and PATH content be merged with what is already there instead of overriding it? Or perhaps this is somehow done automatically? This may be obvious to you, but sadly not me :-/.

As it is, I don't have the time and energy to understand and figure all this out, then test it again etc. So if you want it to use that method I think it makes more sense if someone else (@giordano?) opens a PR with it. Feel free to close this PR then! 

In the meantime, this patch allowed me to use PProf.jl and that in turn allowed me to tune some code substantially using the Memory Allocation Profiler. Thanks !!!

NHDaly commented:
Does https://github.com/JuliaPackaging/Yggdrasil/pull/5652 supersede this PR? Does PProf now work even if you don't have graphviz installed? 🎉 thanks Valentin, btw!

fingolfin commented:
So is #16 now resolved by @vchuravy's https://github.com/JuliaPackaging/Yggdrasil/pull/5652 or not? If it is, the README should be updated and #16 closed.

Just in case it isn't, I've rebased this PR and added a compat bound for Graphviz_jll.

I only bring this up now because I rediscovered this old PR. And I rediscovered it because my student @SoongNoonien was at my office and mentioned difficulties he had with getting PProf to run, because it couldn't find graphviz. Unfortunately I don't have details and he didn't have this laptop there.

giordano commented:
> So is https://github.com/JuliaPerf/PProf.jl/issues/16 now resolved by @vchuravy's https://github.com/JuliaPackaging/Yggdrasil/pull/5652 or not?

I think so because lately I've been using `PProf.jl` without having had to install graphviz manually (didn't even notice the README said you have to do that).

> Unfortunately I don't have details and he didn't have this laptop there.

It'd be good to know the details it investigate that.

SoongNoonien commented:
> It'd be good to know the details it investigate that.

What do you want to now? I have the problem on two machines. Both with recent Arch Linux installations and Julia from the Arch repos.

giordano commented:
A full error message, including stacktrace.

giordano commented:
> and Julia from the Arch repos.

Ah, that's bad: https://wiki.archlinux.org/title/Julia.  Don't do that, see https://julialang.org/downloads/ instead.

SoongNoonien commented:
I finally got around trying this again and indeed using the official Julia build fixed the issue. Thank you!
